ADBDEV-3951: Backport of "Implemented InPlaceUpdate to be used for updates made on non-distribution columns"

KnightMurloc opened the pull request:
Implemented InPlaceUpdate to be used for updates made on non-distribution columns.
Currently, ORCA uses Split-Update for updates on both distribution
and non-distribution columns. With this commit,
ORCA uses an InPlaceUpdate whenever updates are made to
non-distribution columns or non-partition keys, and Split Update 
if any of modified columns are either distribution or partition keys.  
Consider below setup where we are updating
non-distibution column, b in the table foo.

`
create table foo(a int, b int);
explain update foo set b=4;
`
ORCA produces plan with Split and Update nodes

```
 Update
   Output: foo.a, foo.b, (DMLAction), foo.ctid
   ->  Split
         Output: foo.a, foo.b, foo.ctid, foo.gp_segment_id, DMLAction
         ->  Result
               Output: foo.a, foo.b, 4, foo.ctid, foo.gp_segment_id
               ->  Seq Scan on public.foo
                     Output: foo.a, foo.b, foo.ctid, foo.gp_segment_id

```

There is no point in using a Split and Update for this as we are updating a
non-distribution column which do not require any redistribution. This
commit uses an InPlace Update to perform updates on non-distribution
columns like Planner. Below is the new plan produced with this commit.

New Plan
```
 Update
   Output: foo.a, "outer".b, foo.ctid
   ->  Result
         Output: foo.a, 4, foo.ctid, foo.gp_segment_id
         ->  Seq Scan on public.foo
               Output: foo.a, foo.ctid, foo.gp_segment_id

```

greenplum 6 specific changes:
1. Some constructors have been changed because the list of arguments in 6X and 7X
are different.
2. fixed a bug in CParseHandlerPhysicalDML::startElement where preserve_oids_xml
was used instead of fSplit, which could lead to SIGSEGV during DML node parsing.
3. Changed create_index_hot test. Removed disabling the optimizer before updating
since ORCA no longer uses split update in this case.
4. Implement In-Place Update for DML node. Added the exec Update call in the DML
node. Added partitioning selection in case of In-Place update by sorted table.
Added a call to reconstructMatchingTupleSlot in execUpdate to update plans
generated by ORCA to handle cases of dropped attributes.


original commit: https://github.com/greenplum-db/gpdb/commit/3ced85b65732728edff66fd9a2ce6d7485d65a06

Sanath97 committed:
Implemented InPlaceUpdate to be used for updates made on non-distribution columns.
Currently, ORCA uses Split-Update for updates on both distribution
and non-distribution columns. With this commit,
ORCA uses an InPlaceUpdate whenever updates are made to
non-distribution columns or non-partition keys, and Split Update
if any of modified columns are either distribution or partition keys.
Consider below setup where we are updating
non-distibution column, b in the table foo.

`
create table foo(a int, b int);
explain update foo set b=4;
`
ORCA produces plan with Split and Update nodes

```
Update on public.foo
   ->  Result
         Output: foo_1.a, foo_1.b, (DMLAction), foo_1.ctid, foo_1.gp_segment_id
         ->  Split
               Output: foo_1.a, foo_1.b, foo_1.ctid, foo_1.gp_segment_id, DMLAction
               ->  Seq Scan on public.foo foo_1
                     Output: foo_1.a, foo_1.b, 4, foo_1.ctid, foo_1.gp_segment_id
```

There is no point in using a Split and Update for this as we are updating a
non-distribution column which do not require any redistribution. This
commit uses an InPlace Update to perform updates on non-distribution
columns like Planner. Below is the new plan produced with this commit.

New Plan
```
Update on public.foo
   ->  Seq Scan on public.foo foo_1
         Output: foo_1.a, 4, foo_1.ctid, foo_1.gp_segment_id
 Optimizer: Pivotal Optimizer (GPORCA)
```

greenplum 6 specific changes:
1. Some constructors have been changed because the list of arguments in 6X and 7X
   are different.
2. fixed a bug in CParseHandlerPhysicalDML::startElement where preserve_oids_xml
   was used instead of fSplit, which could lead to SIGSEGV during DML node parsing.
3. Changed create_index_hot test. Removed disabling the optimizer before updating
   since ORCA no longer uses split update in this case.

(cherry picked from commit 3ced85b65732728edff66fd9a2ce6d7485d65a06)

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/53331

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/53341

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/53456

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/53506

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/53536

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/53552

RekGRpth commented:
If it's a backport, then maybe you should have cherrypicked it to keep the authorship.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60486

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/890515

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/890516

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/890509

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/890510

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60723

BenderArenadata commented:
Failed job Orca unittests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/905014

BenderArenadata commented:
Failed job ORCA code linter on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/905019

BenderArenadata commented:
Failed job Orca unittests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/905013

RekGRpth commented:
> implement `execUpdate` call for DML node

If it's not a backport, would it be better to make it a separate commit (perhaps a backport of some other commit)?

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/905017

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/905018

KnightMurloc commented:
> > implement `execUpdate` call for DML node
> > not compute actionColIdx for InPlace Update
> 
> If it doesn't apply to the specified backport, then might it be better to make it a separate commit(s) (perhaps a backport(s) of some other commit(s))?

These changes are necessary for the patch to work. GP7 uses the Modify Table node instead of DML. I'm not sure if we want to port these changes.

RekGRpth commented:
> > > implement `execUpdate` call for DML node
> > > not compute actionColIdx for InPlace Update
> > 
> > 
> > If it doesn't apply to the specified backport, then might it be better to make it a separate commit(s) (perhaps a backport(s) of some other commit(s))?
> 
> These changes are necessary for the patch to work. GP7 uses the Modify Table node instead of DML. I'm not sure if we want to port these changes.

I understand that they are needed, I just suggest making them by separate commit(s).

KnightMurloc commented:
But then the authorship of the original commit will be lost. If this option suits us, then I can transfer the 6X-specific changes to a separate commit.

RekGRpth commented:
> But then the authorship of the original commit will be lost. If this option suits us, then I can transfer the 6X-specific changes to a separate commit.

If it is possible, you may create separate commit(s) for 6X-specific changes and then cherry-pick for backport with authorship

KnightMurloc commented:
That is, to put the implementation of the ExecUpdate call for the DML node in a separate commit and make a backport on top of it?

RekGRpth commented:
> That is, to put the implementation of the ExecUpdate call for the DML node in a separate commit and make a backport on top of it?

Yes if it is possible

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/905011

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/905012

KnightMurloc committed:
Implement In-Place Update for DML node

Added the exec Update call in the DML node. Added partitioning selection in case
of In-Place update by sorted table. added a call to reconstructMatchingTupleSlot
in execUpdate to update plans generated by ORCA to handle cases of dropped
attributes.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60806

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/910204

BenderArenadata commented:
Failed job Orca unittests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/910199

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/910205

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60829

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/912054

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/912055

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60874

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/915100

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/915101

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/915095

RekGRpth commented:
Could you

1) Expand the description of the first commit.

2) Add to the description of the second commit that this is a cherry-pick of such and such a commit, and also that such and such was changed for that reason.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61132

BenderArenadata commented:
Failed job ORCA code linter on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/930491

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61145

BenderArenadata commented:
Failed job ORCA code linter on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/930908

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/930906

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/930896

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/930907

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/930898

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61165

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/932547

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/932548

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61167

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/932602

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/932603

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61190

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/933595

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/933597

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/933598

RekGRpth commented on src/test/isolation2/expected/modify_table_data_corrupt_optimizer.out:
```diff
@@ -163,30 +162,29 @@ ABORT
 update tab1 set a = 999 from tab2, tab3 where tab1.a = tab2.a and tab1.b = tab3.a;
-ERROR:  distribution key of the tuple (0, 1) doesn't belong to current segment (actually from seg0) (nodeModifyTable.c:602)  (seg1 127.0.1.1:6003 pid=78344) (nodeModifyTable.c:602)
+UPDATE 1
 abort;
```
Why is it still necessary to have an abortion if it’s successful?

KnightMurloc replied:
we perform an abort not because the query fails, but because we don't want to save the result.

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -116,6 +116,14 @@ ExecDML(DMLState *node)
+		 */
+		if (isUpdate && !isSplit)
+		{
+			node->ps.state->es_result_relation_info = slot_get_partition(node->cleanedUpSlot,node->ps.state);
```
1) line is wider than 80 chars
2) no space before comma
4) braces are redundant

RekGRpth replied:
These comments are still not corrected!

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -57,20 +57,20 @@ ExecDML(DMLState *node)
+	bool isUpdate = node->ps.state->es_plannedstmt->commandType == CMD_UPDATE;
+	bool isSplit = plannode->actionColIdx != 0;
+	// if it's not in place update
+	if (isSplit)
```
```suggestion
	if (AttributeNumberIsValid(plannode->actionColIdx))
```

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -57,20 +57,20 @@ ExecDML(DMLState *node)
-
-	bool isUpdate = false;
-	if (node->ps.state->es_plannedstmt->commandType == CMD_UPDATE)
+	DMLAction action = -1;
```
is -1 allowed to DMLAction (0, 1) ?
```suggestion
	int action = -1;
```

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -116,6 +116,14 @@ ExecDML(DMLState *node)
+		/*
+		 * Perform partition selection for in place update
+		 */
+		if (isUpdate && !isSplit)
```
```suggestion
		if (isUpdate && !AttributeNumberIsValid(plannode->actionColIdx))
```

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -183,6 +191,25 @@ ExecDML(DMLState *node)
 				   PLANGEN_OPTIMIZER /* Plan origin */,
 				   isUpdate);
 	}
+	else /* DML_UPDATE */
```
may be put this in first, as in 7X?
```suggestion
	if (!AttributeNumberIsValid(plannode->actionColIdx))
	...
	else if (DML_INSERT == action)
	...
	else /* DML_DELETE */
	...
```

KnightMurloc replied:
in 7X, this logic is inside the update-only block. whereas the DML node handles other operations as well. Therefore, I see no point in changing the order.

RekGRpth replied:
`action == -1` and `!AttributeNumberIsValid(plannode->actionColIdx)` are equivalent conditions

KnightMurloc replied:
Okay, I get it.

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -183,6 +191,25 @@ ExecDML(DMLState *node)
+				   NULL, //oldtuple
+				   node->cleanedUpSlot,
+				   NULL, //planSlot
+				   &node->mt_epqstate,
```
my be add same to ExecDelete (5th arg)?

RekGRpth commented on src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:
```diff
@@ -3057,6 +3058,89 @@ CExpressionPreprocessor::PexprTransposeSelectAndProject(CMemoryPool *mp,
+	{
+		CColRef *pcrInsert = (*pdrgpcrInsert)[ul];
+		CColRef *pcrDelete = (*pdrgpcrDelete)[ul];
+
```
missed comment
```suggestion
		// Checking if column is either distribution or partition key.
```

RekGRpth commented on src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp:
```diff
@@ -345,14 +346,12 @@ CLogicalDML::OsPrint(IOstream &os) const
 		return COperator::OsPrint(os);
 	}
 
-	os << SzId() << " (";
```
Why was this line removed?

RekGRpth commented on src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp:
```diff
@@ -345,14 +346,12 @@ CLogicalDML::OsPrint(IOstream &os) const
 	CUtils::OsPrintDrgPcr(os, m_pdrgpcrSource);
 	os << "], Action: (";
 	m_pcrAction->OsPrint(os);
-	os << ")";
```
Why was this line removed?

RekGRpth commented on src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp:
```diff
@@ -361,8 +360,47 @@ CLogicalDML::OsPrint(IOstream &os) const
 		os << ", ";
 		m_pcrSegmentId->OsPrint(os);
 	}
-
```
Why was this line removed?

RekGRpth commented on src/backend/gporca/libgpopt/src/xforms/CXformUpdate2DML.cpp:
```diff
@@ -141,30 +151,47 @@ CXformUpdate2DML::Transform(CXformContext *pxfctxt, CXformResult *pxfres,
-		CColRef *pcrInsert = (*pdrgpcrInsert)[ul];
-		CColRef *pcrDelete = (*pdrgpcrDelete)[ul];
-		if (pcrInsert != pcrDelete)
+		const ULONG num_cols = pdrgpcrInsert->Size();
```
Why was this line moved here? In original commit it was not touched.

KnightMurloc replied:
I thought it made sense since its value is only used inside if.

RekGRpth replied:
Yes, but it is backport

RekGRpth replied:
I think we should preserve the original commit code as much as possible. To make it easier to backport other patches that depend on this code.

BenderArenadata commented:
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/937514

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61347

RekGRpth commented on src/include/nodes/plannodes.h:
```diff
@@ -44,7 +44,8 @@ typedef enum PlanGenerator
 	DML_DELETE,
-	DML_INSERT
+	DML_INSERT,
+	DML_INVALID
```
I do not think it's a good idea.

KnightMurloc replied:
we can get away from this if we change the order in if and check for updates first. something like that. what do you think?
```c
if (!AttributeNumberIsValid(plannode->actionColIdx))
{
...
}
else if (DML_INSERT == action)
{
...
}
else /* delete */
{
...
}
```

RekGRpth replied:
> what do you think?

https://github.com/arenadata/gpdb/pull/609#discussion_r1438077428

KnightMurloc replied:
Fine. but won't this slow down the insertion and deletion operations? since at least the insertion will probably happen more often?

RekGRpth replied:
This is exactly what was done in the 7X.

KnightMurloc replied:
in 7X, this is done only for Update whereas we will do it for all DML operations.

RekGRpth replied:
How about
```c
if (DML_INSERT == action)
{
...
}
else if (DML_DELETE == action)
{
...
}
else
{
Assert(!AttributeNumberIsValid(plannode->actionColIdx))
...
}
```

RekGRpth replied:
or
```c
switch (action)
{
case DML_INSERT: ... break;
case DML_DELETE: ... break;
default: Assert(!AttributeNumberIsValid(plannode->actionColIdx)); ... break;
}
```

KnightMurloc replied:
this does not solve the problem. since we still need to assign some value to the `action` variable, otherwise we will get an undefined behavior.

RekGRpth replied:
Yes, we assign -1 before

KnightMurloc replied:
-1 is not a valid DMLAction value. or are you suggesting changing the variable type to int?

RekGRpth replied:
> are you suggesting changing the variable type to int?

Yes, sure! https://github.com/arenadata/gpdb/pull/609#discussion_r1438075745

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/937526

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/937527

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -177,12 +185,31 @@ ExecDML(DMLState *node)
 				   PLANGEN_OPTIMIZER /* Plan origin */,
 				   isUpdate);
 	}
+	else /* DML_UPDATE */
```
What is DML_UPDATE?

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -57,20 +57,20 @@ ExecDML(DMLState *node)
+	if (AttributeNumberIsValid(plannode->actionColIdx))
 	{
-		isUpdate = true;
+		action = (int) DatumGetUInt32(
```
The original code did not contain an explicit type cast.

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/945814

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/937524

RekGRpth commented:
Can you keep only two commits in this patch?

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/937520

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/937521

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61353

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/945977

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/945978

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61361

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/946342

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/946343

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/946337

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/946336

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61460

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/951972

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/951973

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/951966

RekGRpth commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -1370,8 +1371,14 @@ ExecUpdate(ItemPointer tupleid,
 		checkPartitionUpdate(estate, slot, resultRelInfo);
 
+		if (planGen == PLANGEN_OPTIMIZER)
+		{
```
braces are redundant

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61487

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/953686

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/953687

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/953684

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -183,6 +189,25 @@ ExecDML(DMLState *node)
 				   PLANGEN_OPTIMIZER /* Plan origin */,
 				   isUpdate);
 	}
+	else /* UPDATE */
```
May we combine common code of delete and update? like
```c
if (DML_INSERT == action)
{
	// insert code
}
else
{
	// common code
	if (action == DML_DELETE)
	{
		// delete code
	}
	else /* UPDATE */
	{
		// update code
	}
}
```

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -261,6 +286,7 @@ ExecInitDML(DML *node, EState *estate, int eflags)
 	dmlstate->junkfilter = ExecInitJunkFilter(node->plan.targetlist,
 			dmlstate->ps.state->es_result_relation_info->ri_RelationDesc->rd_att->tdhasoid,
 			dmlstate->cleanedUpSlot);
+	estate->es_result_relation_info->ri_junkFilter = dmlstate->junkfilter;
```
Why we need this? May be add some words about it into commit/PR description?

KnightMurloc replied:
it was needed for EPQ logic, but since it was decided to abandon it in this PR, it is no longer needed.

RekGRpth replied:
I suggest to move it to next EPQ patch

RekGRpth commented on src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalDML.h:
```diff
@@ -40,8 +40,6 @@ class CLogicalDML : public CLogical
 	};
 
-	static const WCHAR m_rgwszDml[EdmlSentinel][10];
-
```
original commit does not remove this empty line

RekGRpth commented on src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp:
```diff
@@ -22,9 +22,6 @@
 
-const WCHAR CLogicalDML::m_rgwszDml[EdmlSentinel][10] = {
-	GPOS_WSZ_LIT("Insert"), GPOS_WSZ_LIT("Delete"), GPOS_WSZ_LIT("Update")};
-
```
original commit does not remove this empty line

RekGRpth commented on src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp:
```diff
@@ -346,9 +347,9 @@ CLogicalDML::OsPrint(IOstream &os) const
 
 	os << SzId() << " (";
-	os << m_rgwszDml[m_edmlop] << ", ";
+	CLogicalDML::PrintOperatorType(os, m_edmlop, m_fSplit);
```
```suggestion
	m_ptabdesc->Name().OsPrint(os);
```

RekGRpth commented on src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp:
```diff
@@ -346,9 +347,9 @@ CLogicalDML::OsPrint(IOstream &os) const
 	os << SzId() << " (";
-	os << m_rgwszDml[m_edmlop] << ", ";
+	CLogicalDML::PrintOperatorType(os, m_edmlop, m_fSplit);
+	os << "Affected Columns: [";
```
```suggestion
	CLogicalDML::PrintOperatorType(os, m_edmlop, m_fSplit);
	os << "Affected Columns: [";
```

RekGRpth commented on src/backend/gporca/libgpopt/src/operators/CLogicalDML.cpp:
```diff
@@ -346,9 +347,9 @@ CLogicalDML::OsPrint(IOstream &os) const
-	os << m_rgwszDml[m_edmlop] << ", ";
+	CLogicalDML::PrintOperatorType(os, m_edmlop, m_fSplit);
+	os << "Affected Columns: [";
 	m_ptabdesc->Name().OsPrint(os);
```
```suggestion
```

RekGRpth commented on src/backend/gporca/libgpopt/src/operators/CLogicalUpdate.cpp:
```diff
@@ -285,7 +289,15 @@ CLogicalUpdate::OsPrint(IOstream &os) const
 
```
original patch removes this empty line

RekGRpth commented on src/backend/gporca/libnaucrates/src/parser/CParseHandlerPhysicalDML.cpp:
```diff
@@ -125,6 +126,15 @@ CParseHandlerPhysicalDML::StartElement(const XMLCh *const,	// element_uri,
+			m_parse_handler_mgr->GetDXLMemoryManager(), fSplit,
+			EdxltokenSplitUpdate, EdxltokenPhysicalDMLUpdate);
+	}
+
```
original commit adds two empty lines here

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61589

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/958078

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/958079

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/958072

RekGRpth commented:
error `ERROR: distribution key of the tuple (0, 1) doesn't belong to current segment (actually from seg0) (nodeModifyTable.c:XXX)` in the test does not look like a flaky

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/958073

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61705

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/962859

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -172,16 +178,28 @@ ExecDML(DMLState *node)
-				   PLANGEN_OPTIMIZER /* Plan origin */,
-				   isUpdate);
+		if (DML_DELETE == action)
+			/* Correct tuple count by ignoring deletes when splitting tuples. */
```
I think, it is better to use braces here, or even do something
```c
if (DML_DELETE != action)
{
	ExecUpdate(...);
	return slot;
}
```
and then leave delete block unchanged

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/962861

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -172,16 +178,31 @@ ExecDML(DMLState *node)
+					   NULL /* DestReceiver */,
+					   node->ps.state,
+					   true,
+					   PLANGEN_OPTIMIZER);	
```
```suggestion
					   PLANGEN_OPTIMIZER);
```

RekGRpth commented on src/test/isolation2/expected/modify_table_data_corrupt_optimizer.out:
```diff
@@ -163,26 +162,25 @@ ABORT
-(17 rows)
+ QUERY PLAN                                                                                  
+---------------------------------------------------------------------------------------------
+ Update                                                                                      
```
Can you explain the essence of this test?
Why did the plan change, but execution still results in an error?

KnightMurloc replied:
because in this test we have a tuple that is on the wrong segment. in order to update such a tuple, Explicit Redistribute Motion is required, which is not added by ORCA.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61708

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/962989

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/962990

RekGRpth commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -1571,6 +1576,17 @@ lreplace:;
+					 * act as in RR and upper isolation levels.
+					 */
+					if (!epqstate)
+						ereport(ERROR,
```
Can we move this check to the planning stage in ORCA planner and, if unsuccessful, fallback to the postgres planner?

KnightMurloc replied:
This case only occurs when there is a concurrent update. At the planning stage, we don't know if it will be concurrent or not.

bimboterminator1 commented on src/backend/executor/nodeDML.c:
```diff
@@ -116,6 +115,13 @@ ExecDML(DMLState *node)
+		/*
+		 * Perform partition selection for in place update
+		 */
+		if (isUpdate && !AttributeNumberIsValid(plannode->actionColIdx))
```
We don't need to perform partition selection in case of updating leaf partition. Do we?

RekGRpth commented on src/backend/executor/nodeDML.c:
```diff
@@ -116,6 +115,13 @@ ExecDML(DMLState *node)
+		 * Perform partition selection for in place update
+		 */
+		if (isUpdate && !AttributeNumberIsValid(plannode->actionColIdx))
+			node->ps.state->es_result_relation_info =
```
Do the tests check this code branch?

RekGRpth commented on src/backend/executor/nodeModifyTable.c:
```diff
@@ -1370,8 +1371,12 @@ ExecUpdate(ItemPointer tupleid,
 		checkPartitionUpdate(estate, slot, resultRelInfo);
 
+		if (planGen == PLANGEN_OPTIMIZER)
+			slot = reconstructMatchingTupleSlot(slot, resultRelInfo);
```
Do the tests check this code branch?
